app/services/group_export_service.rb

```diff
@@ -0,0 +1,68 @@
+class GroupExportService
+  RELATIONS = %w[
+    all_users
+    all_events
+    all_notifications
+    all_documents
+    all_reactions
+    memberships
+    membership_requests
+    discussions
+    polls
+    poll_options
+    poll_did_not_votes
+    poll_unsubscriptions
+    outcomes
+    stances
+    stance_choices
+    discussion_readers
+    comments
+  ]
+
+  JSON_PARAMS = { groups: {methods: [:type]},
+                  users:  {except: [:encrypted_password,
+                                    :reset_password_token,
+                                    :email_api_key,
+                                    :reset_password_token,
+                                    :unsubscribe_token] }}.with_indifferent_access.freeze
+
+  def self.export(groups, filename)
+    ids = Hash.new { |hash, key| hash[key] = [] }
+    File.open(filename, 'w') do |file|
+      groups.each do |group|
+        puts_record(group, file, ids)
+        RELATIONS.each do |relation|
+          puts "Exporting: #{relation}"
+          group.send(relation).find_each(batch_size: 20000) do |record|
+            puts_record(record, file, ids)
+          end
+        end
+      end
+    end
+  end
+
+  def self.export_filename_for(group)
+    "tmp/#{DateTime.now.strftime("%Y-%m-%d_%H-%M-%S")}_#{group.name.parameterize}.json"
+  end
+
+  def self.puts_record(record, file, ids)
+    table = record.class.table_name
+    return if ids[table].include?(record.id)
+    ids[table] << record.id
+    file.puts({table: table, record: record.as_json(JSON_PARAMS[table])}.to_json)
+  end
+
+  def self.import(filename)
+    tables = File.open(filename, 'r').map { |line| JSON.parse(line)['table'] }.uniq
```

Comment:
There's some stuff that smells a bit here, but I haven't got a better solution off the top of my head given that (I'm assuming) we don't want to port all of the file into memory. I think json parsing each line of the file N+1 times is going to hurt us, and wonder if there's a way to avoid it.

I wouldn't be opposed to, say, iterating through each line of the file, parsing it, and then,
if the table matches the current table, `klass.new` and append it to an array of records
if the table doesn't match, `klass.import`, flush the existing array, and start a new one with the new table

Then we just iterate through the file once, parsing each line onece, and still maintain support for weirdo input like a stray group in a middle of a big run of discussions or whatnot.

Reply:
PS I believe import will silently fail imports with ids that exist already, which would be the same behaviour as this would exhibit

Reply:
import totally fails if some of the ids already exist. so the check is necessary.

Thanks for helping to try and improve, but i think I'd like to call it as good enough for now. It's the best solution (fastest export for large groups by miles) after quite a few attempts and I'd like to move on.